### DISCUSSION TITLE
How to set the default dialer to be the default phone app?

### DISCUSSION THREAD

**Discussion by Gliting29** [vestrel00/contacts-android#294] [Q&A] [answered]
I have converted my application to default dialer using this package using the code below:

```kotlin
import contacts.ui.util.requestToBeTheDefaultDialerApp

requestToBeTheDefaultDialerApp()
```

I want a function to convert back the default Phone app in mobile as default dialer. Is there any possibility to do so? It would be a great help.

**Comment by vestrel00** [comment, marked as answer]
Hey @Gliting29, this is a good question. I spent a few hours searching the internet for an answer. I could not find one. I also played around with the code inside `requestToBeTheDefaultDialerApp` to see what can be done and found no solution. 

Let's take a look together. The code this library current use to change the default dialer app to the app executing the code is in [DefaultDialerRequest.kt](https://github.com/vestrel00/contacts-android/blob/0.3.0/ui/src/main/java/contacts/ui/util/DefaultDialerRequest.kt#L77-L99),

```kotlin
fun Activity.requestToBeTheDefaultDialerApp() {
    ...
    if (
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
        Build.VERSION.SDK_INT < Build.VERSION_CODES.Q
    ) {
        val intent = Intent(TelecomManager.ACTION_CHANGE_DEFAULT_DIALER)
            .putExtra(TelecomManager.EXTRA_CHANGE_DEFAULT_DIALER_PACKAGE_NAME, packageName)
        startActivityForResult(intent, REQUEST_CODE_SET_AS_DEFAULT_DIALER)
    } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        val roleManager = getSystemService(Activity.ROLE_SERVICE) as RoleManager
        startActivityForResult(
            roleManager.createRequestRoleIntent(RoleManager.ROLE_DIALER),
            REQUEST_CODE_SET_AS_DEFAULT_DIALER
        )
    }
    ...
}
```

In the case of `TelecomManager`, we may be able to set the default phone app to be the default dialer if we know its `packageName`. We have to be able to figure this out dynamically (at runtime). We cannot just hard code it because different Android OEMs may have their own phone app that has a different `packageName`.

Even if we somehow make the case of `TelecomManager` work, it will not work for versions starting with Android Q (Android 10 API level 29). I also see no way of changing the default dialer to anything other than the current application using the `RoleManager`. I think this is intentional as Android made things more strict. It is no longer possible to set the default dialer to an app that is not owned by the app executing the above code.

-------

I will close the issue you created from this; #295. Please feel free to investigate further. If you happen to find a solution, I would greatly appreciate it if you could reopen #295, fork this repo, and create a PR. Thanks and good luck!

**Reply by vestrel00** [reply]
I will mark this as the answer for now.
